app/scripts/lib/channels/duplex.js

```diff
@@ -0,0 +1,185 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+/**
+ * A two way channel. Messages can be sent and received. The channel requires
+ * both a sender and a receiver. A sender and a receiver are the concrete
+ * strategies used to send and receive messages. The decoupling of each
+ * direction allows a channel to e.g., send messages via a CustomEvent and
+ * receive messages via a postMessage (Fx Desktop Sync v1).
+ */
+
+define([
+  'underscore',
+  'lib/promise',
+  'lib/channels/base'
+], function (_, p, BaseChannel) {
+  var DEFAULT_SEND_TIMEOUT_LENGTH_MS = 90 * 1000;
+
+  function OutstandingRequests(options) {
+    this._window = options.window;
+    this._sendTimeoutLength = options.sendTimeoutLength || DEFAULT_SEND_TIMEOUT_LENGTH_MS;
+    this._requests = {};
+  }
+
+  OutstandingRequests.prototype = {
+    add: function (messageId, request) {
+      // remove any old outstanding messages with the same messageId
+      this.remove(messageId);
+
+      request.timeout = this._window.setTimeout(function (command) {
+        this._window.console.error('Response not received for: ' + command);
+      }.bind(this, request.command), this._sendTimeoutLength);
+
+      this._requests[messageId] = request;
+    },
+
+    remove: function (messageId) {
+      var outstanding = this.get(messageId);
+      if (outstanding) {
+        this._window.clearTimeout(outstanding.timeout);
+        delete this._requests[outstanding.messageId];
+      }
+    },
+
+    get: function (messageId) {
+      return this._requests[messageId];
+    },
+
+    clear: function () {
+      for (var messageId in this._requests) {
+        this.remove(this._requests[messageId]);
+      }
+    }
+  };
+
+  function DuplexChannel() {
+  }
+
+  _.extend(DuplexChannel.prototype, new BaseChannel(), {
```

Comment:
The idea is to convert the fx-desktop and web-channel channels to use the duplex channel and remove the postmessage_receiver mixin (terrible idea, see #2235)